### DISCUSSION TITLE
Wrong Sea Pressure calculated

### DISCUSSION THREAD

**Discussion by valdus2** [arendst/Tasmota#22182] [Q&A]
I use a BME 280 sensor and accurately displaying 1018.6 hpa at 9m high with 20.5C.

I have also declared in Tasmota the correct latitude, so it can calculate the constant of gravity.

Tasmota shows 1020.6 Sea Pressure when it should be ~1019.6.

I used the following calculators and they all agree:

http://stacjepogody.waw.pl/kalkulator_cisnien.php

http://www.meteo.tomkii.net/pressure_calc/?alt=9&latitude=38.18&abs=1018.5&temp=20.5&x=10&y=11

https://www.mide.com/air-pressure-at-altitude-calculator

Thank you in advance.

**Comment by s-hadinger** [comment]
I think the code is here:

https://github.com/arendst/Tasmota/blob/7ef93b20fd80760fe5d1b34738195dcf83d0b6c7/tasmota/tasmota_support/support.ino#L956-L974

I suspect this might be due to the imprecision of `FastPrecisePow`.

Could you try to compile with `powf` instead? It will grow the firmware size significantly, but will confirm if that's the cause of the issue. If so, we'll try to find an alternative more precise power function.

```C
  float temp = FastPrecisePow(2.718281828f, (17.67f * t) / (t + 243.5f));
```

**Comment by arendst** [comment]
On the other hand, I have several pressure sensors around the house showing deviations of many hPa so I wouldn't mind a difference of 1 hPa as long as I do not plan to use it as the main height measurement in a plane.

**Comment by valdus2** [comment]
@s-hadinger You mean, changing this:
> ```c
> float temp = FastPrecisePow(2.718281828f, (17.67f * t) / (t + 243.5f));
> ```

to this:
`float temp = FastPrecisePowf(2.718281828f, (17.67f * t) / (t + 243.5f));`



@arendst I'm not talking about deviations of hPa, which of course happens between different sensors,
I'm talking strictly about the conversion inside Tasmota of Pressure to Sea Pressure.
(Whatever hpa to use the coversion-equation seams wrong)
  
Thank you.

**Reply by s-hadinger** [reply]
No. I mean `powf()`

**Reply by valdus2** [reply]
Maybe a little more help please, cause I'm not a C programmer...
What to change exactly?

**Reply by sfromis** [reply]
`powf` instead of `FastPrecisePow`

But the code bit quoted by @s-hadinger was the wrong one, the sea pressure calculation is instead in 
https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_support/support.ino#L1010-L1015

```
float ConvertPressureForSeaLevel(float pressure) {
  if (pressure == 0.0f) {
    return pressure;
  }
  return ConvertPressure((pressure / FastPrecisePowf(1.0f - ((float)Settings->altitude / 44330.0f), 5.255f)) - 21.6f);
}
```

**Reply by sfromis** [reply]
BTW. Those sea pressure calculation can only be an estimate, as local meteorological conditions can vary. And the formula already does not use things like temperature and humidity.

**Reply by valdus2** [reply]
> does not use things like temperature and humidity


Ah, I see.....Nevertheless, I'm going to compile with powf and if any difference, I'll report it here.

So in your bit of code I have to change FastPrecisePowf to powf ?



Thank you.
